### ISSUE TITLE
False positive about type-var variance in callable protocol when used in union type

### ISSUE THREAD

**Issue by Azureblade3808** [microsoft/pylance-release#2172]
## Environment data

-   Language Server version: 2021.12.2
-   OS and version: Windows 10
-   Python version (& distribution if applicable, e.g. Anaconda): 3.8.12 Anaconda

## Sample code

```python
from __future__ import annotations

from collections.abc import Iterable
from typing import Callable, Protocol, TypeVar

_T0 = TypeVar("_T0")
_T1 = TypeVar("_T1")
_T2 = TypeVar("_T2")

class TryMapProtocol(Protocol[_T0, _T1, _T2]):
    def __call__(
        self, func: Callable[[_T0], _T1], iterable: Iterable[_T0], /, default_value: _T2
    ) -> Iterable[_T1 | _T2]:
        ...
```

## Expected behavior

No errors.

## Actual behavior

![image](https://user-images.githubusercontent.com/15389744/145925685-1d5f4bff-ee9e-4e71-84ba-86009e5418dd.png)

## Rationale

If a variable `x` conforms to union type `A0 | B`, `x` either conforms to `A0` or `B`. And when `A0` is a subtype of `A1`, `x` either conforms to `A1` or `B`, therefore `x` conforms to union type `A1 | B`. So we can say that `A` in union type `A | B` is covariant, and obviously the same is true with `B`.

And once we accept this, we can see that both `_T1` and `_T2` are covariant with `Iterable[_T1 | _T2]` which is the result type, so `_T1` and `_T2` should not be contra-variant.

**Comment by Azureblade3808** [comment]
May be relevant with #2095.

**Comment by erictraut** [comment]
I think the current behavior is correct. It's consistent with the error provided by mypy.

**Comment by Azureblade3808** [comment]
I confirmed that MyPy has the same behavior, but I still don't agree with it.

Here is a minimal sample which crashes:
```python
from __future__ import annotations

from random import randint
from typing import Protocol, TypeVar

_T0_contra = TypeVar("_T0_contra", contravariant=True)
_T1_contra = TypeVar("_T1_contra", contravariant=True)

class FP(Protocol[_T0_contra, _T1_contra]):
    def __call__(self, x: _T0_contra, y: _T1_contra, /) -> _T0_contra | _T1_contra:
        ...

def func_object_object(a: object, b: object) -> object:
    # The return type is not even relevant with argument types.
    return object()

def return_fp_object_object() -> FP[object, object]:
    # Type erasing.
    # Safe cast here.
    return func_object_object

def return_fp_float_float() -> FP[float, float]:
    # Type erasing.
    # This is really a dangerous cast, but is considered safe because of contra-variance.
    return return_fp_object_object()  

fp_float_float: FP[float, float] = return_fp_float_float()

fv: float = fp_float_float(0.0, 1.0)  # Real type is `object`.
fv += 1  # Crash here.
```

**Comment by heejaechang** [comment]
```python
_T0_contra = TypeVar("_T0_contra", contravariant=True)
_T1_contra = TypeVar("_T1_contra", contravariant=True)

class FP(Protocol[_T0_contra, _T1_contra]):
    def __call__(self, x: _T0_contra, y: _T1_contra, /) -> _T0_contra | _T1_contra:
```

I think for the example above, _T0_contra being used for both parameter and return type is the problem. not sure why pyright won't flag _T0_contra being used for return type as an error. (that might be a bug)

if _T0_contra is used as just return type, then pyright does flag _T0_contra as an error.

```
	"message": "Type variable \"_T0_contra\" used in generic protocol \"FP\" should be covariant",
```

other static language like C# does flag _T0_contra being used as return type as you can see below when used in both places.
![image](https://user-images.githubusercontent.com/1333179/146064308-0bfe761e-c9b3-4c24-a581-9467078be57c.png)

in C#, "in T" means T is contravariant.

right way to do this in C# is this
```csharp
interface I<in TInput, out TOutput> where TInput: TOutput
{
    TOutput getT(TInput t);
}
```

basically, saying input (parameter) is contravariant, and output (return) is covariant and there is relationship between input and output (input must be same or derived type of output)

but currently, python doesn't let one to create relationship between TypeVars (you can't do bound=_T0_contra), but except that, if you put another type var for return type, it will work as you expects.

```python
from __future__ import annotations

from typing import Protocol, TypeVar

_T0_contra = TypeVar("_T0_contra", contravariant=True)
_T1_contra = TypeVar("_T1_contra", contravariant=True)

_T0_co = TypeVar("_T0_co", covariant=True)
_T1_co = TypeVar("_T1_co", covariant=True)

class FP(Protocol[_T0_contra, _T1_contra, _T0_co, _T1_co]):
    def __call__(self, x: _T0_contra, y: _T1_contra, /) -> _T0_co | _T1_co:
        ...

def func_object_object(a: object, b: object) -> object:
    # The return type is not even relevant with argument types.
    return object()

def return_fp_object_object() -> FP[object, object, object, object]:
    # Type erasing.
    # Safe cast here.
    return func_object_object

def return_fp_float_float() -> FP[float, float, float, float]:
    # Type erasing.
    # This is really a dangerous cast, but is considered safe because of contra-variance.
    return return_fp_object_object()  

fp_float_float: FP[float, float, float, float] = return_fp_float_float()

fv: float = fp_float_float(0.0, 1.0)  # Real type is `object`.
fv += 1

```

pylance will show 

```
	"message": "Expression of type \"FP[object, object, object, object]\" cannot be assigned to return type \"FP[float, float, float, float]\"\n  TypeVar \"_T0_co@FP\" is covariant\n    \"object\" is incompatible with \"float\"",
```

unfortunately, I dont think you can use only 1 typevar to annotate both input (contra) and output (covariant) or use invariant type var for parameter or return type.

**Comment by erictraut** [comment]
Upon further inspection, I agree that these should be invariant, not covariant. The algorithm that's used by both mypy and pyright to determine the proper variance is failing in the same manner in this case because of the union of two type variables. The algorithm involves replacing all but one type variable with "object" and then determining with the resulting specialized protocol is assignable to its unspecialized self and vice versa. In this case, the algorithm is failing because of the ambiguities involved TypeVar matching when a union contains two type variables.

This isn't an easy bug to fix, and it's an extreme edge case, so I don't know if it's worth fixing. I think we can keep it open for now and see if other upvote it or if someone offers a fix in mypy. In the meantime, you can use a ` # type: ignore` as a workaround.

**Comment by Azureblade3808** [comment]
Will it help if the algorithm you mentioned also uses `Never` in addition to `object`?

In my rough imagination, the process would be --

```
class FP(Protocol[_T0, _T1]):
    def __call__(self, x: _T0, y: _T1, /) -> _T0 | _T1:
        ...
```
First, replace `_T1` with `object`.
```
class FP(Protocol[_T0]):
    # Result type is calculated from `_T0 | object`.
    def __call__(self, x: _T0, y: object, /) -> object:
        ...
```
Hmm, it seems that `_T0` should be contra-variant in this case.

Then, replace `_T1` with `Never`.
```
class FP(Protocol[_T0]):
    # Result type is calculated from `_T0 | Never`.
    def __call__(self, x: _T0, y: Never, /) -> _T0:
        ...
```
Oh, `_T0` should be invariant in this case.

As a conclusion, `_T0` should be invariant.

**Comment by heejaechang** [comment]
according to this https://www.python.org/dev/peps/pep-0544/#generic-protocols and https://www.python.org/dev/peps/pep-0544/#overriding-inferred-variance-of-protocol-classes

TypeVar used as return type can't be invariant in generic protocol. 

so, even if Eric fixed the bug that pyright doesn't flag _T0_contra and _T1_contra and show errors, the code itself is still not valid

```
class FP(Protocol[_T0, _T1]):
    def __call__(self, x: _T0, y: _T1, /) -> _T0 | _T1:
        ...
```

one just can't use non covariant typevar in return type.

or you guys might be talking about something else and I completely missed the point...

**Comment by erictraut** [comment]
Type variables used in return types must be covariant if they're not used as input parameters or invariant if they're used as both inputs and returns.

**Comment by heejaechang** [comment]
@erictraut ah.. I see.

**Comment by Azureblade3808** [comment]
@erictraut Would you please point me to the relative code file in Pyright? I have got something in mind and like to see if it could fit in.

**Comment by erictraut** [comment]
It's in checker.ts. Search for "_validateProtocolTypeParamVariance".

**Comment by judej** [comment]
verified fixed in 2021.12.2

**Comment by Azureblade3808** [comment]
This has not been fixed yet. Request for reopening.

**Comment by judej** [comment]
This is fixed in 2022.1.0
